### PR TITLE
[opt](nereids) infer result column name in ctas and query stmt

### PR DESCRIPTION
**Pull request by seawinde**
## Proposed changes

Infer name if it is an expression and doesn't alias artificially when create or select stmt in nereids.
The infer name strategy is the same as #24990

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by seawinde** [comment]
run buildall

**Comment by seawinde** [comment]
run buildall

**Review by zddr — APPROVED** [review]
LGTM

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindExpression.java [hunk 1]
```diff
@@ -564,10 +567,16 @@ protected boolean condition(Rule rule, Plan plan) {
             ),
             RuleType.BINDING_RESULT_SINK.build(
                 unboundResultSink().then(sink -> {
-                    List<NamedExpression> outputExprs = sink.child().getOutput().stream()
-                            .map(NamedExpression.class::cast)
-                            .collect(ImmutableList.toImmutableList());
-                    return new LogicalResultSink<>(outputExprs, sink.child());
+
+                    final ImmutableListMultimap.Builder<ExprId, Integer> exprIdToIndexMapBuilder =
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindExpression.java:571]: not see how to process ctas

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/PlanVisitors.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+/**
+ * This is the facade and factory for plan visitor
+ */
+public class PlanVisitors {
+
+    /**
+     * Infer output column name when it refers an expression and not has an alias manually.
+     */
+    public static class InferPlanOutputAlias extends DefaultPlanVisitor<Void, ImmutableMultimap<ExprId, Integer>> {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/PlanVisitors.java:42]: one vistor in on file under dir visitor

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/PlanVisitors.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        private final List<Slot> currentOutputs;
+        private final List<NamedExpression> finalOutputs;
+
+        public InferPlanOutputAlias(List<Slot> currentOutputs) {
+            this.currentOutputs = currentOutputs;
+            this.finalOutputs = new ArrayList<>(currentOutputs);
+        }
+
+        @Override
+        public Void visit(Plan plan, ImmutableMultimap<ExprId, Integer> context) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/PlanVisitors.java:53]: context is not a good name, use a meaningful name

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/PlanVisitors.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+            List<NamedExpression> projects = plan.getExpressions().stream()
```
[the remark this answers is not in the document]
Comment (seawinde) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/PlanVisitors.java:55]: Want to get all node output expressions and check if it's in the currentExprIdAndIndex Map. if contains then infer the name from the currentExprIdAndIndex.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/PlanVisitors.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+                    .filter(expression -> expression instanceof NamedExpression)
+                    .map(NamedExpression.class::cast)
+                    .collect(Collectors.toList());
+
+            ImmutableSet<ExprId> currentOutputExprIdSet = context.keySet();
+            for (NamedExpression projectItem : projects) {
+                ExprId exprId = projectItem.getExprId();
+                // Infer name when alias child is expression and alias's name is from child
+                if (currentOutputExprIdSet.contains(projectItem.getExprId())
+                        && projectItem instanceof Alias
```
[the remark this answers is not in the document]
Comment (seawinde) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/PlanVisitors.java:65]: Yeah, you are right

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/PlanVisitors.java (hunk 3 above)
Comment (seawinde) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/PlanVisitors.java:53]: OK, rename it to`currentExprIdAndIndexMap`.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/PlanVisitors.java (hunk 2 above)
Comment (seawinde) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/PlanVisitors.java:42]: OK

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindExpression.java (hunk 1 above)
Comment (seawinde) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindExpression.java:571]: Ctas stmt is only supported by legacy, It's already supported in another  pr [Infer the column name when create view if the column is expression](https://github.com/apache/doris/pull/24990), and when cats is supported in nereids future, this is also useful.

**Comment by seawinde** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/PlanVisitors.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+                        && ((Alias) projectItem).isNameFromChild()) {
```
[the remark this answers is not in the document]
Comment (seawinde) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/PlanVisitors.java]: Alias construct logical as following:
`    public Expression visitUnboundAlias(UnboundAlias unboundAlias, CascadesContext context) {
        Expression child = unboundAlias.child().accept(this, context);
        if (unboundAlias.getAlias().isPresent()) {
            return new Alias(child, unboundAlias.getAlias().get());
        } else if (child instanceof NamedExpression) {
            return new Alias(child, ((NamedExpression) child).getName());
        } else {
            return new Alias(child);
        }
    }`

if the alias name is set by child.toSql() we should infer the alias name. `isNameFromChild ` field identify the name is from child. Maybe we should add anther field to record the info. WDYT?

**Comment by seawinde** [comment]
run buildall

**Comment by seawinde** [comment]
run buildall

**Review by morrySnow — APPROVED** [review]

**Pull request merged**